# 成员变量一定是不安全的
成员变量往往是线程不安全的，因为成员变量经常被共享，比如下面的情况
```java
package org.lyflexi.thread.unsafeVariable.memberVariable;  
  
import java.util.ArrayList;  
  
/**  
 * @Author: ly  
 * @Date: 2024/3/14 21:26  
 */public class UnsafeTest {  
    static final int THREAD_NUMBER = 2;  
    static final int LOOP_NUMBER = 200;  
  
    //成员变量是不安全的  
    ArrayList<String> list = new ArrayList<>();  
    public static void main(String[] args) {  
        UnsafeTest test = new UnsafeTest();  
        for (int i = 0; i < THREAD_NUMBER; i++) {  
            new Thread(() -> {  
                test.method1(LOOP_NUMBER);  
            }, "Thread" + (i+1)).start();  
        }  
    }  
    public void method1(int loopNumber) {  
        for (int i = 0; i < loopNumber; i++) {  
            method2();  
            method3();  
        }  
    }  
  
    private void method2() {  
        list.add("1");  
    }  
  
    private void method3() {  
        list.remove(0);  
    }  
}
```
打印信息如下，有一定的概率出现异常
```shell
Exception in thread "Thread1" java.lang.IndexOutOfBoundsException: Index: 0, Size: 1
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.remove(ArrayList.java:498)
	at org.lyflexi.thread.unsafeVariable.memberVariable.UnsafeTest.method3(UnsafeTest.java:35)
	at org.lyflexi.thread.unsafeVariable.memberVariable.UnsafeTest.method1(UnsafeTest.java:26)
	at org.lyflexi.thread.unsafeVariable.memberVariable.UnsafeTest.lambda$main$0(UnsafeTest.java:19)
	at java.lang.Thread.run(Thread.java:750)
```
原因如下
new Thread(() -> {  
list.add("1"); // 时间1. 会让内部 size ++  
list.remove(0); // 时间3. 再次 remove size-- 出现角标越界  
}, "t1").start();  
  
new Thread(() -> {  
list.add("2"); // 时间1（并发发生）. 会让内部 size ++，但由于size的操作非原子性, size 本该是2，但结果可能出现1  
list.remove(0); // 时间2. 第一次 remove 能成功, 这时 size 已经是0  
}, "t2").start();  
![[Pasted image 20240314214516.png]]
# 局部变量不保证100%安全
局部变量往往是线程安全的，但是如果局部变量引用的对象逃离方法的作用范围，并且逃离后创建了内部线程，则有可能不安全

即同时满足两个条件：栈上逃逸+方法中创建了内部线程

下面我们就对局部变量进行线程安全分析
## 安全的局部变量
将成员变量`ArrayList<String> list = new ArrayList<>()`设置为method1的局部变量即可保证安全
```java
package org.lyflexi.thread.unsafeVariable.localVariable;  
  
import java.util.ArrayList;  
  
/**  
 * @Author: ly  
 * @Date: 2024/3/14 21:26  
 */
public class SafeTest {  
    static final int THREAD_NUMBER = 2;  
    static final int LOOP_NUMBER = 200;  
  
  
  
    public static void main(String[] args) {  
        SafeTest test = new SafeTest();  
        for (int i = 0; i < THREAD_NUMBER; i++) {  
            new Thread(() -> {  
                test.method1(LOOP_NUMBER);  
            }, "Thread" + (i+1)).start();  
        }  
    }  
    public final void method1(int loopNumber) {  
        //局部变量是安全的  
        ArrayList<String> list = new ArrayList<>();  
        for (int i = 0; i < loopNumber; i++) {  
            method2(list);  
            method3(list);  
        }  
    }  
  
    public void method2(ArrayList<String> list) {  
        list.add("1");  
    }  
  
    private void method3(ArrayList<String> list) {  
        list.remove(0);  
    }  
}
```
在Java中，方法内部创建的对象通常会被分配到栈上，因为它们的生命周期与方法的生命周期相同，这样可以提高内存分配和回收的效率。然而，如果编译器检测到方法内部创建的对象可能会在方法外部被引用，或者对象的生命周期延长到了方法的结束之后，就会发生栈上逃逸，这时编译器会将该对象在堆上分配内存。虽然出现了栈上逃逸，但由于两个线程的局部变量各自逃逸了互不影响，所以最终线程还是安全的
![[Pasted image 20240314214551.png]]

## 不安全的局部变量-方法内创建子线程
即使list是method的局部变量，但如果我们在method3中开辟了新的线程去执行list.remove(0)，还是会导致线程不安全：
原因是当method3内部线程在执行list.remove(0)，有可能外部线程的list.add("1")还未完全执行结束，因为size++还未完成
```java
package org.lyflexi.thread.unsafeVariable.localVariable;  
  
import java.util.ArrayList;  
  
/**  
 * @Author: ly  
 * @Date: 2024/3/14 21:26  
 */public class UnSafeTest {  
    static final int THREAD_NUMBER = 2;  
    static final int LOOP_NUMBER = 200;  
  
  
  
    public static void main(String[] args) {  
        UnSafeTest test = new UnSafeTest();  
        for (int i = 0; i < THREAD_NUMBER; i++) {  
            new Thread(() -> {  
                test.method1(LOOP_NUMBER);  
            }, "Thread" + (i+1)).start();  
        }  
    }  
    public final void method1(int loopNumber) {  
        //局部变量是安全的  
        ArrayList<String> list = new ArrayList<>();  
        for (int i = 0; i < loopNumber; i++) {  
            method2(list);  
            method3(list);  
        }  
    }  
  
    public void method2(ArrayList<String> list) {  
        list.add("1");  
    }  
  
    //虽然局部变量是安全的，但是在方法的调用过程中，内部创建了子线程共享了成员变量，因此又会出现不安全问题  
    private void method3(ArrayList<String> list) {  
        new Thread(()->{list.remove(0);}).start();  
    }  
}
```
打印信息如下：
```shell
Exception in thread "Thread-393" java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.remove(ArrayList.java:498)
	at org.lyflexi.thread.unsafeVariable.localVariable.UnSafeTest.lambda$method3$1(UnSafeTest.java:38)
	at java.lang.Thread.run(Thread.java:750)

Process finished with exit code 0
```
本质原因是我们在方法内部创建了子线程，无法控制method2和method3的同步执行，所以如果我们修改代码如下，同样是不安全的
```java
    public void method2(ArrayList<String> list) { 
	    new Thread(()->{list.add("1"); }).start(); 
         
    }  
  
    //虽然局部变量是安全的，但是在方法的调用过程中，内部创建了子线程共享了成员变量，因此又会出现不安全问题  
    private void method3(ArrayList<String> list) {  
        list.remove(0);
    }  
```
## 不安全的局部变量-子类重写创建子线程

将局部变量传给了其他方法foo，这发生了栈上逃逸

又因为foo方法abstract，其行为是不确定的被称之为外星方法，内部创建子线程会导致不安全的发生，
```java
public abstract class UnSafeSubclassTest {  
  
    public void bar() {  
        // 是否安全  
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");  
        foo(sdf);  
    }  
  
    public abstract void foo(SimpleDateFormat sdf);  
  
  
    public static void main(String[] args) {  
        new SubClass().bar();  
    }  
  
    static class SubClass extends UnSafeSubclassTest{  
        public void foo(SimpleDateFormat sdf) {  
            String dateStr = "1999-10-11 00:00:00";  
            for (int i = 0; i < 20; i++) {  
                new Thread(() -> {  
                    try {  
                        sdf.parse(dateStr);  
                    } catch (ParseException e) {  
                        e.printStackTrace();  
                    }  
                }).start();  
            }  
        }  
    }  
  
}
```
打印信息如下
```shell
Exception in thread "Thread-7" Exception in thread "Thread-0" Exception in thread "Thread-3" Exception in thread "Thread-8" Exception in thread "Thread-2" Exception in thread "Thread-11" Exception in thread "Thread-1" Exception in thread "Thread-5" java.lang.NumberFormatException: empty String
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at java.text.DigitList.getDouble(DigitList.java:169)
	at java.text.DecimalFormat.parse(DecimalFormat.java:2089)
	at java.text.SimpleDateFormat.subParse(SimpleDateFormat.java:2162)
	at java.text.SimpleDateFormat.parse(SimpleDateFormat.java:1514)
	at java.text.DateFormat.parse(DateFormat.java:364)
	at org.lyflexi.thread.unsafeVariable.localVariable.UnSafeSubclassTest$SubClass.lambda$foo$0(UnSafeSubclassTest.java:33)
	at java.lang.Thread.run(Thread.java:750)
java.lang.NumberFormatException: For input string: ""
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Long.parseLong(Long.java:601)
	at java.lang.Long.parseLong(Long.java:631)
	at java.text.DigitList.getLong(DigitList.java:195)
	at java.text.DecimalFormat.parse(DecimalFormat.java:2084)
	at java.text.SimpleDateFormat.subParse(SimpleDateFormat.java:1869)
	at java.text.SimpleDateFormat.parse(SimpleDateFormat.java:1514)
	at java.text.DateFormat.parse(DateFormat.java:364)
	at org.lyflexi.thread.unsafeVariable.localVariable.UnSafeSubclassTest$SubClass.lambda$foo$0(UnSafeSubclassTest.java:33)
	at java.lang.Thread.run(Thread.java:750)
```
# 直接使用线程安全类

最后，以上只是为了演示使用了ArrayList，其实解决这个问题的方法可以通过同步控制（如使用 `synchronized` 关键字）来保护共享数据的访问，确保在同一时间只有一个线程在操作 `ArrayList` 对象。或者直接使用线程安全的数据结构，比如
- `Vector` 
- `Collections.synchronizedList()` 包装的列表
- `java.util.concurrent` 包中提供的线程安全的数据结构，比如 `CopyOnWriteArrayList`。

但是如果涉及到方法的组合调用，还是会出现线程不安全的现象，比如分析下面代码是否线程安全？
```java
Hashtable table = new Hashtable();  
// 线程1，线程2  
if( table.get("key") == null) {  
    table.put("key", value);  
}
```
我们知道Hashtable的api都是线程安全的，因为查看源码其方法上都加了synchronized，但是上述代码组合在并发场景下有可能出现如下情况：
第一个线程执行完get刚释放synchronized，还没来的及执行put，第二个线程就抢到了synchronized再次get("key") == null
![[Pasted image 20240314222524.png]]
最终导致table.put("key", value)执行了两次，后面线程覆盖了前面线程的值